daml-foundations/daml-tools/daml-cli/DA/Cli/Visual.hs

```diff
@@ -74,47 +75,84 @@ darToWorld manifest pkg = AST.initWorldSelf pkgs pkg
     where
         pkgs = map dalfBytesToPakage (dalfsContent manifest)
 
+tplName :: LF.Template -> T.Text
+tplName LF.Template {..} = head (LF.unTypeConName tplTypeCon)
 
-templateInAction :: Action -> LF.TypeConName
-templateInAction (ACreate  (LF.Qualified _ _ tpl) ) = tpl
-templateInAction (AExercise  (LF.Qualified _ _ tpl) _ ) = tpl
+handleChoiceAndAction :: ChoiceAndAction -> LF.ChoiceName
+handleChoiceAndAction (ChoiceAndAction tpl choice _)
+    | LF.chcName choice == LF.ChoiceName "Create" = LF.ChoiceName $ tplName tpl <> "_Create"
+    | LF.chcName choice == LF.ChoiceName "Archive" = LF.ChoiceName $ tplName tpl <> "_Archive"
+    | otherwise = LF.chcName choice
 
-srcLabel :: (LF.TypeConName, Set.Set Action) -> [(String, String)]
-srcLabel (tc, _) = [ ("shape","none"),("label",DAP.renderPretty tc) ]
+-- Making choiceName is very weird
+handleCreateAndArchive :: TemplateChoiceAction -> [LF.ChoiceName]
+handleCreateAndArchive TemplateChoiceAction {..} =  [createChoice, archiveChoice] ++ map handleChoiceAndAction choiceAndAction
+    where archiveChoice = LF.ChoiceName $ tplName template <> "_Archive"
+          createChoice = LF.ChoiceName $ tplName template <> "_Create"
 
-templatePairs :: (LF.TypeConName, Set.Set Action) -> (LF.TypeConName , (LF.TypeConName , Set.Set Action))
-templatePairs (tc, actions) = (tc , (tc, actions))
+-- This is used to generate the node ids and use as look up table
+choiceNameWithId :: [TemplateChoiceAction] -> Map.Map LF.ChoiceName Int
+choiceNameWithId tplChcActions = Map.fromList $ zip choiceActions [0..]
+  where choiceActions =  concatMap handleCreateAndArchive tplChcActions
 
-actionsForTemplate :: (LF.TypeConName, Set.Set Action) -> [LF.TypeConName]
-actionsForTemplate (_tplCon, actions) = Set.elems $ Set.map templateInAction actions
+nodeIdForChoice ::  Map.Map LF.ChoiceName Int -> LF.ChoiceName -> Int
+nodeIdForChoice lookUpdata chc = case Map.lookup chc lookUpdata of
```

Comment:
Missing camel case in `lookUpdata`. Maybe you can think of a more descriptive name anyway.